src/ocl/convolutionocl.cpp

```diff
@@ -588,12 +639,23 @@ static void ConvForwardCheckNumerics(const Handle& handle,
         return;
     }
 
-    miopen::checkNumericsInput(handle, tensors.xDesc, tensors.x);
-    miopen::checkNumericsInput(handle, tensors.wDesc, tensors.w);
+    bool flag = false;
+
+    flag |= miopen::checkNumericsInput(handle, tensors.xDesc, tensors.x);
+    flag |= miopen::checkNumericsInput(handle, tensors.wDesc, tensors.w);
 
     worker();
 
-    miopen::checkNumericsOutput(handle, tensors.yDesc, tensors.y);
+    flag |= miopen::checkNumericsOutput(handle, tensors.yDesc, tensors.y);
+
+    const char* file_name = miopen::GetStringEnv(MIOPEN_DUMP_TENSOR_PATH{});
+    if(flag && static_cast<bool>(file_name))
+    {
+        std::string file_name_str = file_name;
+        DumpTensorToFileFromDevice(handle, tensors.xDesc, tensors.x, file_name_str + "_x.bin");
+        DumpTensorToFileFromDevice(handle, tensors.wDesc, tensors.w, file_name_str + "_w.bin");
+        DumpTensorToFileFromDevice(handle, tensors.yDesc, tensors.y, file_name_str + "_y.bin");
+    }
```

Comment:
It is definitely worth moving dumping functionality into `checkNumericsOutput()` and `checkNumericsOutput()`. This would enable dumping functionality in all primitives.

Getting `file_name_str` can be moved to a function in check_numerics.cpp

`DumpTensorToFileFromDevice()` should be moved there as well.

Reply:
@atamazov If any of these tensor (`tensors.xDesc`, `tensors.wDesc` and `tensors.yDesc`) have `nan`, we need to dump all three and not just one that has `nan`. If we directly print in `checkNumericsOutput()` and `checkNumericsOutput()` we would  end up dumping only the tensor that has `nan`. 

eg: 
Move dumping functionality into `checkNumericsOutput()` and `checkNumericsOutput()`:

`tensors.xDesc` : has `nan`
`tensors.wDesc` : does not have `nan`
`tensors.yDesc` : does not have `nan`

Only `tensors.xDesc` is dumped. But users might want to see all three tensors and not just one. 

I agree moving `DumpTensorToFileFromDevice()` to `check_numerics.cpp`
I also agree moving ` file_name_str to check_numerics.cpp

I am open for suggestion :smiley:

Reply:
Please notice values 4 and 8 at https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/doc/src/DebugAndLogging.md#numerical-checking.
- When several conv operations fail with NaN/Inf, then only the last triplet of problematic buffers will remain on the disk, because each conv operation with NaN/Inf overwrites existing dumps.
- If the user does not want to abort, then only the last set of problematic buffers will remain in disk. This is because each conv operation with NaN/Inf overwrites existing dumps. These dumps are likely useless. It is the fist set of dumps that is of real interest.
- If the user wants to abort, then dumps won't be written at all.

This must be redesigned.

Reply:
[Proposal]
- Create a class with internal state like `check_failed` flag, addresses of buffers (x, w, y) and more.
- Convert `checkNumericsInput()` and `checkNumericsInput()` into class members.
  - Do not THROW immediately; collect stats instead.
  - These methods should, in addition, register addresses of buffers in the class instance.
  - At the same time I would register strings that identify buffers (`dy`, `w`, `dx`) for the dump filenames.
- Create instance of class wherever you would like to perform checks.
- Call `instance.checkNumerics()` methods.
- At the end, call `instance.TearDown()` or smth alike. It should dump buffers and, if necessary, THROW.
  - Note: Using dtor for this looks attractive, but dtors should never fail (throw).

Reply:
Note that the design is still not MT-safe. If MIOpen is used in multiple threads and failure happens simultaneously (which is often the case!) then the same files will be written from different threads, causing mess. The simplest way to avoid that is encoding thread id in the filenames.

Reply:
If the user wants to continue execution after NaN/Inf detected, and have distinct dumps for each failure, then this is possible. Example design:
- Count failures in some integer member. The member should be static in the class (one counter per process).
- The value of counter should be encoded in the filename when dumping.

All the operations with this counter should be protected (mutex?) to achieve MT-safety.

Encoding of thread id (see previous comment) is still needed to correctly work in multi-process (MP) environment.